### PR TITLE
feat(table): add modular table components

### PR DESCRIPTION
**Pull request by colbr**
[Associated Jira Task](https://wilderworld.atlassian.net/browse/ZAP-39?atlOrigin=eyJpIjoiOTM0NWZiM2M4NzhhNGY2YjljMGY4NzZhOTQwMzk1NTEiLCJwIjoiaiJ9)

-----

## 1. Pull request checklist

- [x] Jira task has been moved to the Code Review column
- [x] Link to Jira task at the top of this PR
- [x] Jira task has a link to this PR
- [x] PR name follows [conventional commit spec](https://www.conventionalcommits.org/en/v1.0.0/)

## 2. PR type
 

Feature

## 3. What is the old behaviour?

We only had `AsyncTable`, which wasn't composable enough for our needs.

## 4. What is the new behaviour?

We now have a bunch of `Table` components which combine to make a properly-styled table.

## 5. Other information

I am pushing this one through so we can iterate quickly, there's still a couple of things that need done:

- We will need to add more tests.
- We will need to add stories.
- We should switch `AsyncTable` to use `Table` internally.

### CODE REVIEW

**Comment by colbr** [comment]
Stylelint failing, I'll fix tomorrow

File: src/components/Table/Column.tsx [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+export const Cell = forwardRef<HTMLTableCellElement, ColumnProps>(({ children, ...props }: ColumnProps, ref) => {
+  return (
+    <ColumnElement element="td" {...props} ref={ref}>
+      {children}
+    </ColumnElement>
+  );
+});
+
+const ColumnElement = forwardRef<HTMLTableDataCellElement | HTMLTableHeaderCellElement, ColumnElementProps>(
```
Comment (domw30) [inline src/components/Table/Column.tsx:42]: Nice - we need to refactor `Button` to use `forwardRef` like this to resolve a console error (I think). Will check that out today by using this as an example

File: src/components/Table/Controls.module.scss [hunk 2]
```diff
@@ -0,0 +1,11 @@
+.Container {
+    display: flex;
+    justify-content: flex-end;
+    gap: 1rem;
+
+    height: 40px;
```
Comment (domw30) [inline src/components/Table/Controls.module.scss:6]: nit: could we switch `px` to `rem`? 

`2.5rem`

File: src/components/Table/Controls.tsx [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    </div>
+  );
+});
+
+/************************
+ * View toggle
+ ***********************/
+
+export enum View {
+  Table = 'table',
```
Comment (domw30) [inline src/components/Table/Controls.tsx:33]: nit: should we use uppercase here for the enum members

File: src/components/Table/Message.tsx [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+const TableMessage = ({ children }: TableMessageProps) => {
+  return <div className={styles.Container}>{children}</div>;
+};
+
+/*************************
+ * Table Status Message
+ ************************/
+
+export enum TableStatus {
+  Loading = 'loading',
```
Comment (domw30) [inline src/components/Table/Message.tsx:25]: nit: uppercase if we decide to change the enum members that have been commented on.

File: src/components/Table/Table.module.scss [hunk 5]
```diff
@@ -0,0 +1,38 @@
+@import '../../styles/colors';
+
+.Table {
+  width: 100%;
+  border-collapse: collapse;
+
+  th {
+    font-size: 0.75rem;
+    font-weight: bold;
+    color: #{$grey-lighter-2};
```
Comment (domw30) [inline src/components/Table/Table.module.scss:10]: nit: is the `#{}` required?

**Review by domw30 — DISMISSED** [review]
Looks good to me @colbr 🚀  Nothing to stop merging

**Comment by domw30** [comment]
I think we also may need to run an install and push to get storybook running mate

Screenshot 2023-02-23 at 10 47 41

File: src/components/Table/Controls.module.scss (hunk 2 above)
Comment (colbr) [reply src/components/Table/Controls.module.scss:6]: Good spot, thanks!

**Comment by colbr** [comment]
> I think we also may need to run an install and push to get storybook running mate

Strange, I can't reproduce this one. Can you try `rm -rf node_modules && yarn && yarn storybook`?

**Comment by domw30** [comment]
> > I think we also may need to run an install and push to get storybook running mate
> 
> Strange, I can't reproduce this one. Can you try `rm -rf node_modules && yarn && yarn storybook`?

Nice, all running fine after removing and re-installing

File: src/components/Table/Table.module.scss (hunk 5 above)
Comment (colbr) [reply src/components/Table/Table.module.scss:10]: Oh, good point, it isn't!

**Pull request merged**